API.md

```diff
@@ -0,0 +1,319 @@
+
+```typescript=
+/* eslint-disable no-undef */
+interface WizardRequestBase {
+  subWizard?: boolean;
+}
+
+export interface EditWizardRequest extends WizardRequestBase {
+  element: Element;
+}
+
+export interface CreateWizardRequest extends WizardRequestBase {
+  parent: Element;
+  tagName: string;
+}
+
+export type WizardRequest = EditWizardRequest | CreateWizardRequest;
+
+type EditWizardEvent = CustomEvent<EditWizardRequest>;
+type CreateWizardEvent = CustomEvent<CreateWizardRequest>;
+export type WizardEvent = EditWizardEvent | CreateWizardEvent;
+
+type CloseWizardEvent = CustomEvent<WizardRequest>;
+
+export function newEditWizardEvent(
+  element: Element,
+  subWizard?: boolean,
+  eventInitDict?: CustomEventInit<Partial<EditWizardRequest>>
+): EditWizardEvent {
+  return new CustomEvent<EditWizardRequest>('oscd-edit-wizard-request', {
+    bubbles: true,
+    composed: true,
+    ...eventInitDict,
+    detail: { element, subWizard, ...eventInitDict?.detail },
+  });
+}
+
+export function newCreateWizardEvent(
+  parent: Element,
+  tagName: string,
+  subWizard?: boolean,
+  eventInitDict?: CustomEventInit<Partial<CreateWizardRequest>>
+): CreateWizardEvent {
+  return new CustomEvent<CreateWizardRequest>('oscd-create-wizard-request', {
+    bubbles: true,
+    composed: true,
+    ...eventInitDict,
+    detail: {
+      parent,
+      tagName,
+      subWizard,
+      ...eventInitDict?.detail,
+    },
+  });
+}
+
+export function newCloseWizardEvent(
+  wizard: WizardRequest,
+  eventInitDict?: CustomEventInit<Partial<WizardRequest>>
+): CloseWizardEvent {
+  return new CustomEvent<WizardRequest>('oscd-close-wizard', {
+    bubbles: true,
+    composed: true,
+    ...eventInitDict,
+    detail: wizard,
+  });
+}
+
+declare global {
+  interface ElementEventMap {
+    ['oscd-edit-wizard-request']: EditWizardRequest;
+    ['oscd-create-wizard-request']: CreateWizardRequest;
+    ['oscd-close-wizard']: WizardEvent;
+  }
+}
+```
+
+## Theming
+
+OpenSCD core sets the following CSS variables on the plugin:
+
+```css
+* {
+  --oscd-primary: var(--oscd-theme-primary, #2aa198);
+  --oscd-secondary: var(--oscd-theme-secondary, #6c71c4);
+  --oscd-error: var(--oscd-theme-error, #dc322f);
+
+  --oscd-base03: var(--oscd-theme-base03, #002b36);
+  --oscd-base02: var(--oscd-theme-base02, #073642);
+  --oscd-base01: var(--oscd-theme-base01, #586e75);
+  --oscd-base00: var(--oscd-theme-base00, #657b83);
+  --oscd-base0: var(--oscd-theme-base0, #839496);
+  --oscd-base1: var(--oscd-theme-base1, #93a1a1);
+  --oscd-base2: var(--oscd-theme-base2, #eee8d5);
+  --oscd-base3: var(--oscd-theme-base3, #fdf6e3);
+
+  --oscd-text-font: var(--oscd-theme-text-font, 'Roboto');
+  --oscd-icon-font: var(--oscd-theme-icon-font, 'Material Icons');
+}
+```
+
+It is expected that the fonts `--oscd-theme-text-font` and `--oscd-theme-icon-font` will be loaded in OpenSCD's `index.html` file. OpenSCD core does not load any fonts by itself.
+
+## Missing
```

Comment:
The proposal in [this comment](https://github.com/OpenEnergyTools/scl-lib/issues/54#issuecomment-2031855291) is something we could include in the API defined here. What do you think about that?

To me it looks like a great fit for the new `SetAttributes` API, since we already provide automatically generated unique namespace prefixes there (as opposed to the old `Update` API that left prefixes to the user). Collating all namespace definitions for newly inserted elements and newly set attributes into the root node is quite a bit of extra effort on the API implementor's side, but would make the end user's experience much smoother.

It would also be the first bit of logic in OpenSCD core that is `SCL` specific. At that point, we could also consider reintroducing built-in `getReference` (for `reference === undefined` in `Insert`) by letting OpenSCD core depend on `scl-lib`.

What do you think about these points, @danyill and @JakobVogelsang ?

Reply:
I would love to have it in. That is from what I understand an implementation issue in the core component. The proposed API allows to do that, right?

Reply:
I like it. It offloads responsibility onto core for some slightly complicated semantics and helps keep the SCL documents consistent with the API and that's a good "core task".

Reply:
Could one of you send a pull request onto my `add-api-docs` branch with a draft of the required API changes? If possible, including the automatic `reference` if undefined, since that makes keeping the document at least schema-valid a much easier task for plugin authors who don't want to think about the standard too much.

Reply:
We've decided in today's meeting to not implement those features within core, but to leave `getReference` in `scl-lib` and add a functional API for automatic handling of extension namespaces there instead.